### PR TITLE
validate client version

### PR DESCRIPTION
**Pull request by honnix**
this will give server side a chance to reject request from certain
client version.

### CODE REVIEW

File: styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ public CliResource(String schedulerServiceBaseUrl, Storage storage) {
             "GET", BASE + "/events/<cid>/<eid>/<iid>",
             rc -> eventsForWorkflowInstance(arg("cid", rc), arg("eid", rc), arg("iid", rc))))
 
-        .map(r -> r.withMiddleware(Middleware::syncToAsync))
+        .map(r -> r.withMiddleware(h -> Middleware.syncToAsync(h)))
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java]: is this change necessary?

Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/CliResource.java]: not really, didn't realise I changed it!

File: styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java [hunk 2]
```diff
@@ -214,4 +214,9 @@ public void storeResource(Resource resource) throws IOException {
   public Optional<Backfill> backfill(String id) throws IOException {
     return datastoreStorage.getBackfill(id);
   }
+
+  @Override
+  public List<String> clientBlacklist() {
+    return datastoreStorage.clientBlacklist();
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java]: Is it likely that we'll need to update this list at runtime? If not, could it just be a "hardcoded" list either in code or the config file?

Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java]: I think we will need to update this list at runtime. There was a case that we released a faulty CLI but the server side had no problem.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 3]
```diff
@@ -363,7 +365,9 @@ private String apiUrl(CharSequence... parts) {
   }
 
   private ByteString send(Request request) throws ExecutionException, InterruptedException {
-    final Response<ByteString> response = client.send(request).toCompletableFuture().get();
+    final Response<ByteString> response =
+        client.send(request.withHeader("Client-Version", STYX_CLI_VERSION)).toCompletableFuture()
```
Comment (danielnorberg) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: Should we include "Styx" in the header? E.g. `Styx-Client-Version: Styx CLI 1.2.3`. 

Alternatively, would it perhaps make sense to use the `User-Agent` header? E.g. `User-Agent: Styx CLI 1.2.3`.

File: styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java [hunk 4]
```diff
@@ -18,10 +18,9 @@
  * -/-/-
  */
 
-package com.spotify.styx;
+package com.spotify.styx.util;
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java]: Would it make sense to use the Guava memoizing supplier instead of rolling our own?

http://google.github.io/guava/releases/21.0/api/docs/com/google/common/base/Suppliers.html#memoizeWithExpiration-com.google.common.base.Supplier-long-java.util.concurrent.TimeUnit-

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 3 above)
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: Would be better to use ```User-Agent``` indeed. I wasn't sure how Apollo will deal with this header, hopefully not rewriting it. I will do some experiment or read doc.

File: styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java (hunk 2 above)
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java]: Ok, I guess we could just deploy a new version of the server then that blacklists that client, but I don't have a very strong opinion.

**Comment by danielnorberg** [comment]
This PR seems to include PR https://github.com/spotify/styx/pull/72, might be good to rebase?

Overall LGTM apart from some comments 👍

File: styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java (hunk 4 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java]: Something from Guava sounds more reliable. What do you think, @rouzwawi ? Did we have some problem then decided to have our own cached solution?

**Comment by honnix** [comment]
You are right, I must have messed up with branches.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 3 above)
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: It looks at least Apollo doesn't touch ```User-Agent```, so use that instead of reinventing the wheel.

File: styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java (hunk 4 above)
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java]: This was just a product of my ignorance :)

I think it does exactly the same as the guava momoized supplier. Feel free to remove it.

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java [hunk 5]
```diff
@@ -596,9 +599,19 @@ private Entity backfillToEntity(Backfill backfill) {
...
         .filter(e -> e.contains(PROPERTY_CONFIG_CONCURRENCY))
         .map(e -> e.getLong(PROPERTY_CONFIG_CONCURRENCY));
   }
+
+  List<String> clientBlacklist() {
+    final String blacklist = readConfigString(PROPERTY_CONFIG_CLIENT_BLACKLIST, "");
+    if (blacklist.isEmpty()) {
+      return ImmutableList.of();
+    } else {
+      return Arrays.stream(blacklist.split(";")).map(
```
Comment (rouzwawi) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Datastore supports lists of values natively. That's probably better than manual splitting on `;`.

File: styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java [hunk 6]
```diff
@@ -37,6 +44,32 @@
  */
 public final class Middlewares {
 
+  private static class ClientValidator implements
+                               Middleware<AsyncHandler<? extends Response<?>>,
+                                   AsyncHandler<? extends Response<ByteString>>> {
+    private final Supplier<List<String>> clientBlacklist;
+
+    ClientValidator(Storage storage) {
+      clientBlacklist = new CachedSupplier<>(storage::clientBlacklist, Instant::now);
```
Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java]: I would rather pass in a `Supplier<List<String>>` as a dependency instead of constructing it from `Storage` here.

This should also use the injected `Time` instance instead of using `Instant::now`. But this should be solved by just moving the construction up to the main service class.

File: styx-api-service/src/main/java/com/spotify/styx/StyxApi.java [hunk 7]
```diff
@@ -100,13 +102,16 @@ public void create(Environment environment) {
     final StyxConfigResource styxConfigResource = new StyxConfigResource(storage);
     final CliResource cliResource = new CliResource(schedulerServiceBaseUrl, storage);
 
+    final Middleware clientValidator = Middlewares.clientValidator(storage);
```
Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: Pass in `new CachedSupplier<>(storage::clientBlacklist, time)` instead. See my other comment.

Comment (rouzwawi) [inline styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: I guess there's one difference in that `CachedSupplier` supports injecting time while the Guava decorator uses wall clock time.

Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: Let's keep it then? Feels like still worth it.

File: styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java (hunk 6 above)
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java]: Good point. I like this better. Didn't really like the class anyway. We can get rid of it and use closure instead.

File: styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java (hunk 5 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java]: Will fix.

File: styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java (hunk 2 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/storage/AggregateStorage.java]: If you don't have strong opinion, I would prefer runtime configuration. ;) hehehe

File: styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java (hunk 4 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java]: We keep it as it offers a bit more than guava.

**Comment by honnix** [comment]
@danielnorberg have time to take a look again?

**Comment by honnix** [comment]
@rouzwawi can I get a thumb up?

File: styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java [hunk 8]
```diff
@@ -67,4 +73,21 @@ private Middlewares() {
       }
     });
   }
+
+  public static Middleware<AsyncHandler<? extends Response<?>>,
+      AsyncHandler<? extends Response<ByteString>>> clientValidator(
+      Supplier<Optional<List<String>>> supplier) {
+    return innerHandler -> requestContext -> {
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java]: Add a metric for the number of rejected requests?

**Comment by danielnorberg** [comment]
👍

File: styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java (hunk 8 above)
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/Middlewares.java]: We don't have any metric emitted from styx-api-service. This feels like a separated PR. I will add todo for now.
